pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -24,130 +24,181 @@ import (
+		"v1alpha2": v1alpha2NodePrepareResouce,
+		"v1alpha3": v1alpha3NodePrepareResouce,
+	}
+
+	nodeUnprepareResourceHandlers = map[string]nodeUnprepareResourceHandler{
+		"v1alpha2": v1alpha2NodeUnprepareResouce,
+		"v1alpha3": v1alpha3NodeUnprepareResouce,
+	}
+)
 
 func NewDRAPluginClient(pluginName string) (drapb.NodeClient, error) {
 	if pluginName == "" {
 		return nil, fmt.Errorf("plugin name is empty")
 	}
 
-	existingPlugin := draPlugins.Get(pluginName)
+	existingPlugin := draPlugins.get(pluginName)
 	if existingPlugin == nil {
 		return nil, fmt.Errorf("plugin name %s not found in the list of registered DRA plugins", pluginName)
 	}
 
-	return &draPluginClient{
-		pluginName: pluginName,
-		plugin:     existingPlugin,
-	}, nil
+	return existingPlugin, nil
 }
 
-func (r *draPluginClient) NodePrepareResources(
+func (p *plugin) NodePrepareResources(
 	ctx context.Context,
 	req *drapb.NodePrepareResourcesRequest,
 	opts ...grpc.CallOption,
-) (resp *drapb.NodePrepareResourcesResponse, err error) {
+) (*drapb.NodePrepareResourcesResponse, error) {
 	logger := klog.FromContext(ctx)
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
-	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
 
-	conn, err := r.plugin.getOrCreateGRPCConn()
+	conn, err := p.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	nodeClient := drapb.NewNodeClient(conn)
-	nodeClientOld := drapbv1alpha2.NewNodeClient(conn)
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
-	resp, err = nodeClient.NodePrepareResources(ctx, req)
-	if err != nil {
-		status, _ := grpcstatus.FromError(err)
-		if status.Code() == grpccodes.Unimplemented {
-			// Fall back to the older gRPC API.
-			resp = &drapb.NodePrepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodePrepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				respOld, errOld := nodeClientOld.NodePrepareResource(ctx,
-					&drapbv1alpha2.NodePrepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodePrepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				} else {
-					result.CDIDevices = respOld.CdiDevices
-				}
-				resp.Claims[claim.Uid] = result
-			}
-		}
+	p.RLock()
+	fallbackVersion := p.fallbackVersion
+	p.RUnlock()
+
+	handler, ok := nodePrepareResourceHandlers[fallbackVersion]
+	if !ok {
+		return nil, fmt.Errorf("unsupported plugin version: %s", fallbackVersion)
 	}
 
-	return
+	response, err := handler(ctx, conn, p, req)
+	logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", response, "err", err)
+	return response, err
 }
 
-func (r *draPluginClient) NodeUnprepareResources(
+func (p *plugin) NodeUnprepareResources(
```

Comment:
This function looks similar to `NodePrepareResources`. The only difference is `req` type. Would it make sense to use generics here?

Reply:
hmm, let me try